File: airplane/Plane.java
package airplane;
import java.util.Random;

public class Plane {
    private String description;
    private String title;
    private int PassengerCapacity;
    private CargoBay cargoBay;
    private EquipmentCompartment[] equipmentCompartments;
    private PassengerCompartment[] passengerCompartments;
    private Random rand = new Random();
    private int N = (rand.nextInt() & Integer.MAX_VALUE) % 4 + 2;               //number of passenger compartments

    Plane(){
        title = "Boeing 787 Dreamliner";
        description = "This is a plane";
        PassengerCapacity = 200;
        cargoBay = new CargoBay();
        equipmentCompartments = new EquipmentCompartment[3];
        for(int i = 0; i<3; i++){
            equipmentCompartments[i] = new EquipmentCompartment();
        }
        passengerCompartments = new PassengerCompartment[N];
        for(int i = 0; i<N; i++){
            passengerCompartments[i] = new PassengerCompartment(true);
        }
        System.out.println("Plane created!");

    }

    public String getDescription() {
        return description;
    }
    public String getTitle(){
        return title;
    }
    public int getPassengerCapacity() {
        return PassengerCapacity;
    }
    public CargoBay getCargoBay() {
        return cargoBay;
    }
    public EquipmentCompartment[] getEquipmentCompartments() {
        return equipmentCompartments;
    }
    public PassengerCompartment[] getPassengerCompartments() {
        return passengerCompartments;
    }

    public int getN(){
        return N;
    }

    public void ready_check(){
        boolean ECR = true, PCR = true;
        for(int i = 0; i<3; i++){
            if(!equipmentCompartments[i].ready_check())
                ECR = false;
        }
        for(int i = 0; i<N; i++){
            if(!passengerCompartments[i].ready_check())
                PCR = false;
        }
        if(cargoBay.ready_check() && ECR && PCR) {
            System.out.println("Plane ready to take off!!!");
        }
        else{
            System.out.println("Plane not ready.");
        }

    }

    public void process(SecurityEmployee se){
        cargoBay.process(se);
        for(int i = 0; i<3; i++){
            equipmentCompartments[i].process(se);
        }
        for(int i = 0; i<N; i++){
            passengerCompartments[i].process(se);
        }
    }

    public void process(MaintenanceEmployee me){
        cargoBay.process(me);
        for(int i = 0; i<3; i++){
            equipmentCompartments[i].process(me);
        }
    }

    public void process(CleaningEmployee ce){
        cargoBay.process(ce);
        for(int i = 0; i<N; i++){
            passengerCompartments[i].process(ce);
        }
    }

}

File: askisi3/Plane.cpp
#include "Plane.h"
#include <iostream>
using namespace std;

Plane::Plane()
: N(2), PassengerCapacity(200), PC(NULL), description("This is a plane"), title("Boeing 707"), EC(NULL) {

    cout << "Plane created!" << endl;
    PC = new PassengerCompartment*[N];
    for(int i = 0; i<N; i++){
        PC[i] = new PassengerCompartment(true);
    }
    EC = new EquipmentCompartment*[3];
    for(int i = 0; i<3; i++) {
        EC[i] = new EquipmentCompartment;
    }
}

Plane::~Plane() {
    for (int i = 0; i < N; ++i)
    {
        delete PC[i];
    }
    delete[] PC;
    for (int i = 0; i < 3; ++i)
    {
        delete EC[i];
    }
    delete[] EC;
    cout << "Plane about to be destroyed!" << endl;
}

String Plane::getTitle() const {
    return title;
}

String Plane::getDesc() const {
    return description;
}

int Plane::getPCapacity() const {
    return PassengerCapacity;
}

CargoBay Plane::getCargoB() const{
    return CargoB;
}

EquipmentCompartment* Plane::getEC() const{
    return *EC;
}

PassengerCompartment* Plane::getPCompartment() const {
    return *PC;
}

bool Plane::ready_check() const{
    bool PCR = true, ECR = true;    //PCR-ECR stand for Passenger-Equipment Compartment Ready
    for(int i = 0; i<N; i++){
        if(!PC[i]->ready_check()) {
            PCR = false;
        }
    }
    for(int i = 0; i<3; i++){
        if(!EC[i]->ready_check()) {
            ECR = false;
        }
    }
    return CargoB.ready_check() && ECR && PCR;
}

void Plane::process(SecurityEmployee &SE) {
    CargoB.process(SE);
    for(int i = 0; i<N; i++){
        PC[i]->process(SE);
    }
    for(int i = 0; i<3; i++) {
        EC[i]->process(SE);
    }
}

void Plane::process(MaintenanceEmployee &ME) {
    CargoB.process(ME);
    for(int i = 0; i<3; i++) {
        EC[i]->process(ME);
    }
}

void Plane::process(CleaningEmployee &CE) {
    CargoB.process(CE);
    for(int i = 0; i<N; i++){
        PC[i]->process(CE);
    }
}

bool Plane::equal(Object *ob) const {
    Plane * p = dynamic_cast<Plane *>(ob);
    if(p==NULL)
        return false;
    else
        return (p->id == this->id && p->description.get() == this->description.get() && p->title.get() == this->title.get() && p->PassengerCapacity == this->PassengerCapacity);
}

Object* Plane::clone() const {
    Plane *p;
    p = new Plane(*this);

    p->EC = new EquipmentCompartment*[3];           //Deep Copy
    for(int i = 0; i<N; i++){
        p->EC[i] = new EquipmentCompartment(*this->EC[i]);
    }

    p->PC = new PassengerCompartment*[N];
    for(int i = 0; i<N; i++){
        p->PC[i] = new PassengerCompartment(*this->PC[i]);
    }
    return p;
}

String Plane::toString() const {
    return String(to_string(id) + title.get() + description.get() + to_string(PassengerCapacity));
}